Xenial   domainpermissions

balcy opened the pull request:
This PR adds 2 new features:
1) domain settings
fixes https://github.com/ubports/morph-browser/issues/154
fixes https://github.com/ubports/morph-browser/issues/147
fixes https://github.com/ubports/morph-browser/issues/195

2) domain permissions
fixes https://github.com/ubports/morph-browser/issues/200

both are added for morph-browser and webapps

balcy committed:
add new sqlite database for domain specific settings

balcy committed:
move DomainSettingsModel to webbrowser-common to make it available for webapps

balcy committed:
move onNavigationRequested to Browser.qml
extend DomainSettingsModel

balcy committed:
add initializations

balcy committed:
correct column names

balcy committed:
correct column

balcy committed:
adapt DomainSettingsModel: populateFromDatabase() instead of fetchMore()
the read methods now access the model data (no extra database queries)

balcy committed:
add (not finished) settings page for domain specific settings (browser and webapps)
make add / remove functions of DomainSettingsModel available in QML
move BrowserPage to the common folder, so that it can be used both for the browser and webapps

balcy committed:
add AllowCustomSchemesDialog
the user can select if the custom url scheme setting should be saved permanently (it is never saved for incognito mode)
navigationRequestDelegate for webapps was called twice on each request, remove one of the connections

balcy committed:
after the permission for custom schemes is given, navigate to the respective url again

balcy committed:
add selectMode to DomainSettingsPage
use SortFilterModel (sort by domain without subdomain) to DomainSettingsPage
add domainWithoutSubdomain() to domain-settings-model
correct DownloadsPage

balcy committed:
- custom url schemes: ask user if current scheme is "file" as well
- adapt GeolocationPermission dialog so that the user can save the permission permanently
- save domain specific zoom factor (includes correction to the save default zoom dialog)
- add default zoom factor to DomainSettingsModel (for database cleanup of entries with default zoom factor)

balcy committed:
sort domainsettings in the backend, save default zoom factor in DomainSettingsPage

balcy committed:
add DomainSettingsSortedModel

balcy committed:
add target to CMakeLists.txt

balcy committed:
correct DomainSettingsSortedModel

balcy committed:
add UserAgentsModel, userAgent -> userAgentId in DomainSettingsModel

balcy committed:
change connection name for domain-settings-user-agents-model

balcy committed:
continue with UserAgentModel, add to DomainSettingsPage

balcy committed:
add ui for custom user agents, adapt / correct models
adapt DomainSettingsPage

balcy committed:
add custom user agents handling for browser and webapps

balcy committed:
prototyping DomainPermissions, enable settings for webapps

balcy committed:
implement domain permissions UI (browser and webcontainer)

balcy committed:
add RequestedByDomain to DomainPermissionsModel
add methods to reset the databases for DomainSettings / DomainPermissions

balcy committed:
display the requesting domain in the domainPermissions page for non-mainframe requests

balcy committed:
add reset function for DomainPermissions / DomainSettings
add incognito flag to DomainPermissions
switch order: first handle custom schemes, than domain permissions

balcy committed:
some corrections / UI improvements for DomainPermissions / DomainSettings

balcy committed:
add lastRequested (DATETIME) to DomainPermissionsModel, to allow sorting by date later

balcy committed:
domainpermissions can be sorted alphabetically or chronologically
delete cache for morph-browser / webapps
add settings page link to chrome of webapps

balcy committed:
rename openLinkInWindowRequested -> openLinkInNewWindowRequested
in incognito mode links to be opened in a new incoginto window

balcy committed:
correct paths for cache deletion, add imports for common lib (FileOperations was moved there)

balcy committed:
update .pot file

balcy committed:
correct domain-utils for IP addresses
try to avoid several confirmations of the same domain in AllowOrBlockDomainDialog

balcy committed:
lastPartOfDomain still contains the dot

balcy committed:
DomainPermissionsModel.setRequestedByDomain is only called for non-mainframe-requests, if they have a different domain

balcy committed:
remove additional setRequestedByDomain call

balcy commented:
hints for testing:

domain settings:
- navigate to https://t.me/ubports/176342
- in the dialog, tick the checkbox "save permanently" and press OK
- the telegram app opens
- in Settings -> Domain Specific Settings "t.me" with the tick "allow custom schemes" is present
- restart morph and visit the URL again, the telegram app opens automatically

zoom:
- navigate to some page, e.g. the initial page
- select "Zoom" from the main menu
- set the zoom to a value that is not the default zoom
- in Settings -> Domain Specific Settings the domain entry with the selected zoom appears
- restart morph and visit the page again.
- select "Zoom" from the main menu, the zoom is the one selected last time, and the page zoom matches that value

location:
- activate location access on the phone
- navigate to a page asking your location, e.g. https://www.w3schools.com/html/tryit.asp?filename=tryhtml5_geolocation
- in this example click on "Try it"
- in the Permission Request, check the "Save permanently" checkbox, and press "Allow"
- your current location is displayed in the box at the bottom of the page
- in Settings -> Domain Specific Settings the entry "www.w3schools.com" (or the page you used) exists and the checkbox "allow location access" is checked
- restart morph and visit the page again
- after pressing "try it" the page does not ask for the location permission this time, and your location is displayed at the bottom of the page

... custom user agents will follow soon

balcy committed:
set index of listview to -1 before adding new user agent

balcy commented:
~~to test the custom user agents:~~
* ~~navigate to https://zn6kx.codesandbox.io/ (user agent page 1) and save it to the favorites~~
* ~~navigate to https://xcdqd.codesandbox.io/ (user agent page 2) and save it to the favorites~~
* ~~for both pages the default (desktop or mobile) user agent should appear~~
* ~~go to Settings -> Domain Specific Settings -> Custom User Agents~~
* ~~click "+", type "u1" as name for the new user agent, and then click "OK"~~
* ~~in the textarea type "dummy user agent 1 ..."~~
* ~~click "+", type "u2" as name for the new user agent, and then click "OK"~~
* ~~in the textarea type "dummy user agent 2 ..."~~
* ~~go to Settings -> Domain Specific Settings~~
* ~~click "+" and type "zn6kx.codesandbox.io" to add the domain (alternatively, change the zoom on the page as described above to add it automatically)~~
* ~~check the box "custom user agent"~~
* ~~click on "u1"~~
* ~~click "+" and type "xcdqd.codesandbox.io" to add the domain (alternatively, change the zoom on the page as described above to add it automatically)~~
* ~~check the box "custom user agent"~~
* ~~click on "u2"~~
* ~~navigate to https://zn6kx.codesandbox.io/~~
* ~~you should see "dummy user agent 1 ..."~~
* ~~navigate to https://xcdqd.codesandbox.io/~~
* ~~you should see "dummy user agent 2 ..."~~

**note:** there seems to be one issue with the page history (back and forward): by changing the user agent it seems to be lost, maybe because it is saved in the profile and assumed to stay constant per webview by QtWebEngine, or it is a problem of how the user agent is changed in morph, not sure here...

balcy committed:
deactivate custom user agents for now

balcy commented:
I've deactivated the custom user agents for now, because I've not found a way to reliably change them for specific domains.
An alternative to changing httpUserAgent of the WebEngineProfile might be https://doc.qt.io/qt-5/qquickwebengineprofile.html#setUrlRequestInterceptor, see https://github.com/qutebrowser/qutebrowser/issues/3636

UniversalSuperBox commented:
Because of the size of this PR, I think merging it will unfortunately need to be held until after the release of OTA-10. All the review and testing I can do might not reveal how the changes interact with each other, and we're a little close to release. There are also some string changes and the translators might hit me with their devices if I merge this so late. :P

UniversalSuperBox commented:
Wow, how close we were to release indeed :roll_eyes:

Review starts *now*!

UniversalSuperBox commented on src/Morph/Web/MorphWebView.qml:
```diff
@@ -322,6 +322,9 @@ WebEngineView {
     readonly property bool lastLoadFailed: internal.lastLoadRequestStatus === WebEngineLoadRequest.LoadFailedStatus
     readonly property string lastLoadRequestErrorString: internal.lastLoadRequestErrorString
     onLoadingChanged: {
+        if (loadRequest.errorCode === 420) {
```
Is this something thrown internally by QtWebEngine or do you use a lot of Spring?

balcy replied:
yes it is internally thrown by QtWebEngine 5.11:
https://github.com/qt/qtwebengine/blob/5.11/src/core/web_contents_delegate_qt.cpp
lines 598 and 600.
But they are no longer using it for 5.13
Ah now I get it: spring framework, no never used that : )
If the custom schemes are not yet allowed, you see a flickering (always reloading) page of the second screenshot of https://github.com/ubports/morph-browser/issues/195
So we can probably remove this again with QtWebEngine 5.13.

UniversalSuperBox replied:
Sounds good

UniversalSuperBox commented on src/app/AllowCustomSchemesDialog.qml:
```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 ubports
```
We are currently unable to have copyright assigned to us. You can keep your copyright if you want.

balcy replied:
could we use "ubports contributors" instead ?

UniversalSuperBox replied:
No, just yourself

balcy replied:
there are examples of "the UBports project" though:
https://github.com/search?p=4&q=org%3Aubports+%22The+UBports+project%22&type=Code

balcy replied:
is it OK if I remove all lines "Copyright 2019 ubports" altogether, and put no copyright to the single files at all ?

UniversalSuperBox replied:
Maybe if the whole project was that way, but in this case there's a convention to follow.

balcy replied:
> We are currently unable to have copyright assigned to us. You can keep your copyright if you want.

ok but the convention was "Copyright 2013-2017 Canonical Ltd." and I don't want to keep the copyright

balcy replied:
I could just write my name for now, and later if the foundation is ready, change it back maybe ?

UniversalSuperBox commented on src/app/AllowCustomSchemesDialog.qml:
```diff
@@ -0,0 +1,61 @@
+
+    message: i18n.tr("The following url with a custom scheme was blocked:") + "\n" +
+                                           url + "\n\n" +
+                                           i18n.tr("Should all custom URL schemes from domain %1 be allowed?".arg(domain));
```
This spacing seems random. Why so far to the right?

balcy replied:
no reason, I'll remove the additional spaces.

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+                onTriggered: {
+                    var promptDialog = PopupUtils.open(Qt.resolvedUrl("PromptDialog.qml"), domainPermissionsPage);
+                    promptDialog.title = i18n.tr("Add domain")
+                    promptDialog.message = i18n.tr("Enter the name of the domain, e.g. example.com (subdomains will be removed).")
```
May this dialog also have a "Current page" option, to add domain settings for the site you're currently viewing?

balcy replied:
sounds good, I think the settings page cannot directly access this information though. (we might need to add another property to the settings page)
for the zoom, location and custom url schemes there is already a workflow that adds the current page as needed. (e.g. if user changes the zoom or the page needs location access)

balcy replied:
this feature I would rather skip for now, as soon as custom user agents will (hopefully) be activated again, it will become handy to have it.
For now the location / custom schemes / zoom do already have a workflow to add the current page automatically.

UniversalSuperBox replied:
Ah, don't worry about it for now then.

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+                            else {
+                                DomainPermissionsModel.insertEntry(domain, false);
+                                reload(domain);
+                            }
```
Wait, what are we doing if the user enters nothing? Does `domain` exist here?

balcy replied:
there is the check `if (text !== "") {` in line 119

UniversalSuperBox replied:
Ah, okay.

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+                visible: !selectMode
+
+                onTriggered: {
+                    var promptDialog = PopupUtils.open(Qt.resolvedUrl("PromptDialog.qml"), domainPermissionsPage);
```
Does `PromptDialog` allow giving the text box a URL type? It seems strange to have it listed as a normal text box. Also makes autocorrect go nuts.

balcy replied:
yeah that is a good point, I try adding that property to the prompt dialog.

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+                readonly property bool isCurrentItem: item.ListView.isCurrentItem
+                readonly property string domain: model.domain
+                height: isCurrentItem ? layout.height : units.gu(5)
+                color: isCurrentItem ? theme.palette.selected.base : theme.palette.normal.background
```
`base` is quite dark now. I suspect it is not the color you want.

balcy replied:
yes I think that was a trade-off where I tried to make it look acceptable with light and dark theme, corrections welcome : )

UniversalSuperBox commented on src/app/DomainSettingsPage.qml:
```diff
@@ -0,0 +1,390 @@
+                            visible: item.ListView.isCurrentItem
+
+                            Label  {
+                                text: i18n.tr("allow custom schemes")
```
"Allow this site to launch other apps on your device"

balcy replied:
was a bit long, have used "allowed to launch other apps"

UniversalSuperBox commented on src/app/DomainSettingsPage.qml:
```diff
@@ -0,0 +1,390 @@
+                            visible: item.ListView.isCurrentItem
+
+                            Label  {
+                                text: i18n.tr("allow location access")
```
"Allow this site to access your location"

balcy replied:
was a bit long, have changed it to "allowed to access your location"

UniversalSuperBox commented on src/app/DomainSettingsPage.qml:
```diff
@@ -0,0 +1,390 @@
+
+                            Label  {
+                                text: i18n.tr("allow location access")
+                            }
```
Is it appropriate to have location settings here? I think that "Settings -> Privacy and Permissions" already contains camera and microphone permissions, so it should also have these.

balcy replied:
I think it was done to have all domain specific settings in one place. (e.g. adding, editing, and removing them), as they are in a common data model.
would it help to put the whole category "domain settings" to "privacy and permissions" ?

balcy replied:
b.t.w. the microphone and camera permissions are not functional at the moment

UniversalSuperBox replied:
Moving them would make sense. A little unfortunate about mic and camera.

UniversalSuperBox commented on src/app/webbrowser/SettingsPage.qml:
```diff
@@ -268,6 +284,36 @@ FocusScope {
+                            objectName: "setDomainWhiteListMode"
+
+                            ListItemLayout {
+                                title.text: i18n.tr("Domain whitelist mode")
```
"Only allow browsing to whitelisted websites"

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+            wrapMode: Text.Wrap
+            width: parent.width
+            horizontalAlignment: Text.AlignHCenter
+            text: i18n.tr("No domain specific permissions available")
```
"No sites have been granted special permissions"

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+
+                        ColumnLayout {
+                            visible: item.ListView.isCurrentItem
+                            CustomizedRadioButton {
```
I'm unable to select any of these radio buttons upon creating a site. I created `example.com` while on a different site, then tried to select a radio button. No dice.

balcy replied:
uh, that is strange, can it be because I am on the edge channel ? Is the "Not Set" radio button selected after you create "example.com" ? For me if I click on the radio buttons or the texts "Blocked" or "Whitelisted", the selection changes.

cibersheep replied:
It happens only if: 
- there is no previous information stored (like used for the first time) 
- go to settings / domain permisions
- add a url
- tap on select

balcy replied:
I removed the database (/home/phablet/.local/share/morph-browser/domainpermissions.sqlite) to have the situation of the first use.
I see the following after adding the url:

![image](https://user-images.githubusercontent.com/848115/64798074-af98c400-d582-11e9-9c72-beb493f7c112.png)

what is "tap on select" referring to ?

cibersheep replied:
Sorry. You might want to read «Tap on select» as «Tap on edit» :D 
Mmm... The menu I see is different, I might not have your latest changes :)

balcy replied:
@UniversalSuperBox @cibersheep  unfortunately I cannot reproduce that, could you maybe add a screenshot,or does the log say sth?
If you activate the checkbox "domain whitelist mode" (old version) / "only allow browseing to whitelisted websites" (new version), does morph properly ask for each new domain if it should be blocked or whitelisted ? Are these settings reflected in the domain permissions menu ?
[edit] with Aquaris 4.5 and Desktop the radio boxes do work for me

UniversalSuperBox replied:
Hmm. I originally reproduced that on my Oneplus One across a few restarts of the browser and phone, but I can't do it any longer. Guess we'll keep watch for that one.

cibersheep replied:
It looks I cannot reproduce it anymore. I guess after rebooting a couple of times is gone. Mmmm....

UniversalSuperBox commented on src/app/DomainSettingsPage.qml:
```diff
@@ -0,0 +1,390 @@
+                readonly property string domain: model.domain
+                readonly property int userAgentId: model.userAgentId
+                height: isCurrentItem ? layout.height : units.gu(5)
+                color: isCurrentItem ? theme.palette.selected.base : theme.palette.normal.background
```
Again, I think that `base` is much darker than you expect.

balcy replied:
yeah might need tweaking

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+
+                            CustomizedRadioButton {
+                                checked: (model.permission === DomainPermissionsModel.Whitelisted)
+                                text: i18n.tr("Whitelisted")
```
"Always allow access"

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+                            }
+                            CustomizedRadioButton {
+                                checked: (model.permission === DomainPermissionsModel.Blocked)
+                                text: i18n.tr("Blocked")
```
"Never allow access"

UniversalSuperBox commented on src/app/DomainPermissionsPage.qml:
```diff
@@ -0,0 +1,290 @@
+        anchors.fill: parent
+        focus: true
+
+        title: i18n.tr("Domain specific permissions")
```
It's unclear what a domain specific permission is. Is "Website Blocking" more accurate?

UniversalSuperBox replied:
Or is "Domain blacklist/whitelist" better

balcy replied:
agreed, maybe the 2nd one.

UniversalSuperBox reviewed (changes requested):
Here's what I've found so far. I believe we have some good bones for functionality here, but it's unclear what some of the settings do.

Paging @cibersheep to try this out too (`sudo ubports-qa install xenial_-_domainpermissions`). There's a lot of new settings, menus, and buttons to play with.

cibersheep commented:
> Paging @cibersheep to try this out too (`sudo ubports-qa install xenial_-_domainpermissions`). There's a lot of new settings, menus, and buttons to play with.

Luvely. I'll have something to do at the start of the week :)

cibersheep commented:
Installed. Were I should look for the changes?

UniversalSuperBox commented:
Tap through Settings a bit. There are *lots* of new options.

I think there are some dialogs to be hit when browsing sites with bad behavior, but I haven't seen them yet.

giiba commented:
I did the install as noted above and like Cibersheep I see no new options... are we sure that instal method works? I had to temporarily mount the apt archive in userdata to make it work so I am confident something was installed but none of the changes seem to be in effect.

cibersheep commented:
Weird. After installing, I see no changes in the Settings menu

UniversalSuperBox commented:
I've triggered a new build in Jenkins. After it finishes, run `sud ubports-qa update` to get the new goodies.

giiba commented:
That did the trick. This is wonderful and solves a few of the major gripes I had with Morph!

At first pass things seem to be working well, nice job.

cibersheep commented:
Nice. Domain Permissions settings needs a little tweak but I think is good for now.

aka: 
- more space between option «allow...» with check boxes
- checkboxes to be aligned to the right

balcy committed:
corrections / change requests:
- adapt label texts
- move domain specific settings to privacy section of settings
- align checkboxes of domain specific settings
- prompt dialogs with url mode and without text prediction

balcy committed:
adapt label for DomainPermissionsPage without entries

balcy committed:
replace color theme.palette.selected.base

balcy committed:
adapt copyright of new files

UniversalSuperBox commented on src/app/AllowCustomSchemesDialog.qml:
```diff
@@ -0,0 +1,61 @@
+/*
```
Hmm. Functionally, this dialog is fine. But I wonder if we can move it around and change the text to make it clearer. @cibersheep, do you have any thoughts?

![screenshot20190915_174200521](https://user-images.githubusercontent.com/21966173/64928869-89259380-d7e3-11e9-907f-823f346f0d45.png)

cibersheep replied:
Yep. I think is a bit confusing because of the structure.
I got there doing:
 - going to t.me[blbalba]
- Allow/Block dialog t.me. Tapped allow
- Custom URL schemes dialog saying «t.me is blocked»

In this Dialog, what about something like this?:

**Custom URL Scheme**
This site is trying to ¿open? the URL scheme: `tg:[blablab]`

This will try to dispatch this url to a system app if possible. Do you want to proceed?

Always allow URL schemes from t.me []

[allow]
[Cancel]

cibersheep replied:
So, first I get 

![imatge](https://user-images.githubusercontent.com/6640041/64931833-55765800-d83b-11e9-8cef-2ec37ef3fffd.png)

then 

![imatge](https://user-images.githubusercontent.com/6640041/64931840-6626ce00-d83b-11e9-9e2e-599cd9d769a0.png)

UniversalSuperBox replied:
For reference, here are some examples of the same dialog from Chromium and Firefox:

![image](https://user-images.githubusercontent.com/21966173/64971377-3a244080-d86d-11e9-9034-653ba799a8c8.png)

![Screenshot at 2019-09-16 10-30-23](https://user-images.githubusercontent.com/21966173/64971392-427c7b80-d86d-11e9-8399-9ccd7461899e.png)

Of course we can't select the application to launch, nor would the browser know which application is going to be launched.

cibersheep replied:
Firefox message seems more understandable (+1 FF)

cibersheep commented:
About the design tweaks. I have some thoughts:

On the Allow/Block Dialog

![imatge](https://user-images.githubusercontent.com/6640041/64931909-d46b9080-d83b-11e9-9722-7e24e77b3b13.png)

I would choose a main action and color only that one.

**Block Domain**
About to access to domain XXX. Do you want to block this domain from now on?

[Block domain] (in red)
[Allow] (uncolored)
[Cancel] (uncolored)

Or the opposite, if you think the main action is to allow domains_

[Allow] (in green)
[Block domain] (uncolored)
[Cancel] (uncolored)

On Dialogs like this:
![imatge](https://user-images.githubusercontent.com/6640041/64932003-4cd25180-d83c-11e9-94d9-50c4ed49e146.png)

I would show a specific text in the button. More like:
**This is happening**
Do you want to [action]?
[Action]
[Cancel]

And reserve the "Ok" answer for informative texts:
**This happened**
This is what happened right now.
[Ok]

In the Domain Specific Settings:

![imatge](https://user-images.githubusercontent.com/6640041/64932132-48f2ff00-d83d-11e9-8493-403655bafda0.png)

- I would go for a clear grey (Silk, even Porcelain)
- Center the check boxes to the right with ore space (ListItems even)

In the Domain Specific Permissions:

![imatge](https://user-images.githubusercontent.com/6640041/64932206-a9823c00-d83d-11e9-8fcc-8409a53dac6d.png)

- Same with grey
- I would show Not set, Blocked and Whitelisted in `theme.palette.normal.backgroundText`
- I would use an exposed [combo button](https://api-docs.ubports.com/sdk/apps/qml/Ubuntu.Components/ComboButton.html) to show the options
- And use the Suru icons for the allowed and blocked icons :)

If you want me to help with any / all those, ping me

balcy commented:
- the background color has been adapted (commit 3 days ago)
  (the color is derived from the background, but relatively lighter or darker, depending on the theme, feel free to change / suggest the factor here)
- the alignment of the checkboxes has been made (commit 3 days ago)
- the custom url scheme dialog has been adapted (current commit)
  [the text for the checkbox does not include the domain, because it could be too long]
- the blacklist/whitelist page has been adapted, so that the radio box labels are no longer bold and no longer green / red

for the "allow block" dialog I think neither allow, nor block is the default action:
Depending on the page and what the user wants, they have to take the decision.
And both selections result in a blacklist or whitelist entry in the database. (the same page might include content from desired and non-desired domains)

- I think radio boxes for an enum with only 3 fixed values are easier to use than combo boxes
- with the suru icons I'm not sure which one could be better, I've used color overlays (green / red) to make the list easier to read.

balcy committed:
adapt custom schemes dialog (rearrange content), change label styles of DomainPermissionsPage

cibersheep commented:
Coool... let me try your changes and I get back to you.

Meanwhile:
- Suru Icons you might use the ones named: "ok" for allowed, "cancel" for blocked. And they can be colored as you wish
- About the Allow/Deny Dialog... I agree that is tricky. What about following the Restart menu style?
Meaning: only one of the options is colored (red or green accordingly) the user still needs to choose (and not forcing block or allow as I suggested)

![imatge](https://user-images.githubusercontent.com/6640041/65031645-e6375b80-d941-11e9-9f5a-71b4239ba824.png)

- Radio buttons. The thing is I can't remember any where in the system we use them. We use ListItems for that:

ie.
GSM provider:
![imatge](https://user-images.githubusercontent.com/6640041/65031974-8d1bf780-d942-11e9-879c-d6a2ff5a39d5.png)

Lock screen after x seconds:
![imatge](https://user-images.githubusercontent.com/6640041/65032057-b6d51e80-d942-11e9-8d6b-63031eb90c84.png)

balcy commented:
- icons: ah they really look much better, especially in the dark mode (would you keep dialog-question-symbolic for "NotSet" ?

- Allow / Deny dialog: Let's say the default would be allow (green), because most pages the user navigates to would be allowed. But if "block" is grey then, the user might not know that he will no longer be able to access the page from then.
If we really have to avoid having a "positive" and "negative" button in one dialog, I'd prefer "Allow" grey, and "Block" red. But I like the current version most, because the colors reflect the status in the permissions page.

- the radio boxes: the examples from the settings page are selected items from a list, but here we select an option of a single list item.
While we already highlight the list item (the domain), the "sublist" item would need another color again to see the selcted item. Also, the selected item would have a tick on the right, that looks like the symbol to whitelisted items. I would still vote for those here, if many users complain we can change them : )

balcy committed:
change domain permissions icons

cibersheep commented:
Thank you for your work, I have tested the updates you did :)

- The info icon for Not set could work until I can design something specific (if you like)
- For the Allow / Deny Dialog, I know that is a bit special but I really think we should be quite meticulous with the core apps therefore I would go with the top as a red block button
- And for the radio buttons: I see what you mean. What about leaving it like it is for now and come back to it in a later revision?

balcy committed:
adapt AllowOrBlockDomainDialog : main action block, other buttons without color
add title to AlertDialog for blocked domains

balcy commented:
>     * The info icon for Not set could work until I can design something specific (if you like)
ok thanks, for now I'm fine with the icon
> 
>     * For the Allow / Deny Dialog, I know that is a bit special but I really think we should be quite meticulous with the core apps therefore I would go with the top as a red block button
ok I tried to change that
>     * And for the radio buttons: I see what you mean. What about leaving it like it is for now and come back to it in a later revision?
it is not carved in stone : )

balcy commented:
@UniversalSuperBox  ok ready for merge ?

UniversalSuperBox commented on src/app/GeolocationPermissionRequest.qml:
```diff
@@ -23,29 +23,37 @@ import Ubuntu.Components.Popups 1.3
-        onClicked: reject()
+    ListItemLayout {
+        visible: showAllowPermanentlyCheckBox
+        title.text: i18n.tr("Save permanently")
```
Since this was changed in the custom URL scheme request, it should also be changed to "Remember decision" here.

UniversalSuperBox commented on src/app/AllowCustomSchemesDialog.qml:
```diff
@@ -0,0 +1,61 @@
+         }
+    }
+    Button {
+        text: i18n.tr("Allow")
```
I noticed that allowing or denying one of these is remembered until the end of the session. That is, if I deny a URL open without checking "Remember decision" then click the URL again, I'm not asked to perform the action again.

Now, this is good for not spamming the user. Is it what you expected to happen?

balcy replied:
yes this is done on purpose. Since the for geolocation requests this behavior (remembering the decision for the current session) is already done by the backend (QtWebEngine), I wanted the same behavior here.

By the way: in incognito mode the "remember decision" checkboxes are not available.

UniversalSuperBox replied:
Okay, just making sure.

UniversalSuperBox commented on src/app/AllowOrBlockDomainDialog.qml:
```diff
@@ -0,0 +1,64 @@
+
+    property string domain
+
+    message: i18n.tr("About to access domain %1. Do you want to block this domain from now on?".arg(domain));
```
This text still seems unclear to me. As a user, I don't know what will happen if I choose each option. Since this dialog only appears when whitelist-only mode is enabled (I think), let's refine it as "You're trying to access *domain* but it is not on your approved domain whitelist. Would you like to continue to block the domain, allow browsing to the domain, or go back?"

UniversalSuperBox replied:
Maybe that's too long.

balcy replied:
yes it is right, the dialog does only appear in "whitelist-only" mode.

There are two types of requests though: mainFrame requests (navigating to a different page) and requests within a page to another domain.

For the non-mainFrame requests I would suggest sth. like: 
"_parentDomain_ is trying to access _domain_ but it is not on your approved domain whitelist. Would you like to continue to block the domain, allow loading data from the domain, or just ignore this request ?"

UniversalSuperBox replied:
How would I see the second type of request?

balcy replied:
e.g https://nasa.gov
those requests are async, so several dialogs will appear. But the order the user blocks them does not matter.

cibersheep replied:
Mmm... tested the latest changes. I think the experience could be polished a bit but I don't think is bad how it is now. How about going ahead how it is and revisit later on?

balcy replied:
by the way: the allow / block dialog was inspired from the "Jade Diamond" browser : )
![image](https://user-images.githubusercontent.com/848115/65726808-3d61cc80-e0b6-11e9-9848-f584803d1c01.png)

cibersheep replied:
Jade Diamond is not a core app  ;)

balcy committed:
adapt text on AllowOrBlockDomainDialog and GeolocationPermissionRequest

balcy committed:
update pot file

balcy committed:
correct typo

UniversalSuperBox reviewed (approved):
Here we go.

giiba commented:
Nice work folks!

I've been testing the webapp and site specific zoom settings and it is really nice to have as it solves some of the problems that required more complicted solutions for zoom settings.

One curiosity: is there a way to apply zoom settings earlier in the rendering process? At the moment the pages are rendered then displayed, and then everything shifts afterwards as the content is fit to the zoom setting. This makes for a slightly awkward user experience as after a page loads things start moving before settling down.

giiba commented:
One thing I discovered this morning:

If a website asks for permission to location information the dialog won't let me disallow and remember the choice. If I check the 'remember decision' option the deny button is disabled.

Is there a reason for this? There are sites that ask it every page load, and I would like to just block the request permanently, as opposed to having to choose every time.

![screenshot20190930_111828902](https://user-images.githubusercontent.com/7121141/65905017-98284c00-e374-11e9-806e-c76030ae6566.png)

balcy commented:
yeah good point. At that time it was implemented like a permission that can be revoked later (just a boolean flag "allow location access" in the database).
So permanent blocking was not thought of there, and would require a new field in the database.
[edit: actually "allow permanently" was renamed to "remember decision" later]

giiba commented:
Well it doesn't hinder the functionality, as Morph(stable) asks every time anyways on a site like that. Really the problem is the website.

balcy commented:
> One curiosity: is there a way to apply zoom settings earlier in the rendering process? At the moment the pages are rendered then displayed, and then everything shifts afterwards as the content is fit to the zoom setting. This makes for a slightly awkward user experience as after a page loads things start moving before settling down.

it might be possible to use WebEngineLoadRequest.LoadStartedStatus instead of WebEngineLoadRequest.LoadSucceededStatus, but we have to make sure that we still detect whenever the current domain changes. (WebViewImpl.qml, 879)

giiba commented:
I played a bit with WebEngineView.LoadStartedStatus but it doesn't seem to trigger. If the current LoadSucceededStatus is replaced the zoom is never applied, and if I add a fourth 'if' under onLoadingChanged for LoadStartedStatus (and copy+paste from LoadSucceededStatus) there is no difference in behaviour.

Might be futile, I don't know enough, and it really isn't a problem.
